File: Samples/AppAmbit.App.ObjC/models/CmsExampleModel.m
#import "CmsExampleModel.h"

@implementation AuthorRelation

- (instancetype)initWithDictionary:(NSDictionary *)dict {
    self = [super init];
    if (self && dict && [dict isKindOfClass:[NSDictionary class]]) {
        _author = dict[@"author"];
    }
    return self;
}

- (NSString *)displayString {
    if (self.author && ![self.author isKindOfClass:[NSNull class]]) return self.author;
    return @"Unknown Author";
}

@end

@implementation CmsExampleModel

- (id)safeObject:(id)obj {
    return (obj == [NSNull null]) ? nil : obj;
}

- (instancetype)initWithDictionary:(NSDictionary *)dict {
    self = [super init];
    if (self && [dict isKindOfClass:[NSDictionary class]]) {
        _id = [self safeObject:dict[@"id"]];
        _title = [self safeObject:dict[@"title"]];
        _body = [self safeObject:dict[@"body"]];
        id catObj = [self safeObject:dict[@"category"]];
        if ([catObj isKindOfClass:[NSArray class]]) {
            _category = catObj;
        } else if ([catObj isKindOfClass:[NSString class]] && [(NSString *)catObj length] > 0) {
            _category = @[catObj];
        } else {
            _category = nil;
        }
        _featuredImage = [self safeObject:dict[@"featured_image"]];
        
        id authorObj = [self safeObject:dict[@"author"]];
        if ([authorObj isKindOfClass:[NSDictionary class]]) {
            _author = [[AuthorRelation alloc] initWithDictionary:authorObj];
        } else {
            _author = nil;
        }
        
        id vwObj = [self safeObject:dict[@"views_count"]];
        _viewsCount = vwObj ? [vwObj doubleValue] : 0.0;
        
        id pubObj = [self safeObject:dict[@"is_published"]];
        _isPublished = pubObj ? [pubObj boolValue] : NO;
        
        _eventDate = [self safeObject:dict[@"event_date"]];
        _scheduledPublishAt = [self safeObject:dict[@"scheduled_publish_at"]];
        _authorEmail = [self safeObject:dict[@"author_email"]];
        _metaData = [self safeObject:dict[@"meta_data"]];
    }
    return self;
}

@end

File: Samples/AppAmbit.App.Swift/models/CmsExampleModel.swift
import Foundation
import AppAmbit

struct AuthorRelation: Decodable {
    let author: String?

    init(from decoder: Decoder) throws {
        if let single = try? decoder.singleValueContainer(), let str = try? single.decode(String.self) {
            author = str
        } else {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            author = try container.decodeIfPresent(String.self, forKey: .author)
        }
    }

    private enum CodingKeys: String, CodingKey { case author }

    var displayString: String {
        return author ?? "Unknown Author"
    }
}

struct CmsExampleModel: Decodable, Identifiable {
    let id: String?
    let title: String?
    let body: String?
    let category: [String]?
    let author: AuthorRelation?
    let featuredImage: String?
    
    let viewsCount: Double?
    let isPublished: Bool?
    let eventDate: String?
    let scheduledPublishAt: String?
    let authorEmail: String?
    let metaData: AnyDecodable?

    enum CodingKeys: String, CodingKey {
        case id, title, body, category, author
        case featuredImage = "featured_image_url"
        
        case viewsCount = "views_count"
        case isPublished = "is_published"
        case eventDate = "event_date"
        case scheduledPublishAt = "scheduled_publish_at"
        case authorEmail = "author_email"
        case metaData = "meta_data"
    }

    var uid: String { id ?? UUID().uuidString }
}
